File: crates/spfs/spenv/cli/_cmd_info.py
from typing import Union
import os
import argparse

from colorama import Fore, Style

import spenv


def register(sub_parsers: argparse._SubParsersAction) -> None:

    info_cmd = sub_parsers.add_parser("info", help=_info.__doc__)
    info_cmd.add_argument(
        "--verbose", "-v", action="count", help="increase the verbosity of output"
    )
    info_cmd.add_argument("refs", metavar="REF", nargs="*")
    info_cmd.set_defaults(func=_info)


def _info(args: argparse.Namespace) -> None:
    """Display information about the current environment or specific items."""

    config = spenv.get_config()
    repo = config.get_repository()
    if not args.refs:
        _print_global_info()
        return
    for ref in args.refs:
        item = repo.read_object(ref)
        _pretty_print_ref(item, args.verbose)


def _pretty_print_ref(obj: spenv.storage.Object, verbosity: int = 0) -> None:

    if isinstance(obj, spenv.storage.Platform):
        print(f"{Fore.GREEN}platform:{Fore.RESET}")
        print(f" {Fore.BLUE}refs:{Fore.RESET} " + spenv.io.format_digest(obj.digest))
        print(f" {Fore.BLUE}stack:{Fore.RESET}")
        for ref in obj.stack:
            print(f"  - " + spenv.io.format_digest(ref))

    elif isinstance(obj, spenv.storage.Layer):
        print(f"{Fore.GREEN}layer:{Fore.RESET}")
        print(f" {Fore.BLUE}refs:{Fore.RESET} " + spenv.io.format_digest(obj.digest))
        print(f" {Fore.BLUE}manifest:{Fore.RESET}")
        if verbosity == 0:
            max_entries = 10
        if verbosity == 1:
            max_entries = 50
        else:
            max_entries = 0

        count = 0
        for path, entry in obj.manifest.walk():
            print(f"  {entry.mode:06o} {entry.kind.value} {path}")
            count += 1
            if max_entries and count == max_entries:
                print(f" {Style.DIM}  ...[truncated] use -vv for more{Style.RESET_ALL}")

    else:
        print(repr(obj))


def _print_global_info() -> None:
    """Display the status of the current runtime."""

    config = spenv.get_config()
    repo = config.get_repository()

    runtime = spenv.active_runtime()
    if runtime is None:
        print(f"{Fore.RED}No Active Runtime{Fore.RESET}")
        return

    print(f"{Fore.GREEN}Active Runtime:{Fore.RESET}")
    print(f" {Fore.BLUE}id:{Fore.RESET} {runtime.ref}")
    print(f" {Fore.BLUE}stack:{Fore.RESET}")
    stack = runtime.get_stack()
    for ref in stack:
        print(f"  - " + spenv.io.format_digest(ref))
    print()

    if not runtime.is_dirty():
        print(f"{Fore.RED}No Active Changes{Fore.RESET}")
        return

    print(f"{Fore.BLUE}Active Changes:{Fore.RESET}")
    base = spenv.compute_runtime_manifest(runtime)
    current = spenv.tracking.compute_manifest("/env")
    diffs = spenv.tracking.compute_diff(base, current)
    print(spenv.io.format_changes(diffs))

File: crates/spfs-cli/common/src/cmd_info.rs
// Copyright (c) Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use clap::Args;
use colored::*;

use spfs::io::{self, DigestFormat};
use spfs::{self, prelude::*};

/// Display information about the current environment, or specific items
#[derive(Debug, Args)]
pub struct CmdInfo {
    /// Operate on a remote repository instead of the local one
    ///
    /// This is really only helpful if you are providing a specific ref to look up.
    #[clap(long, short)]
    remote: Option<String>,

    /// Tag or id to show information about
    #[clap(value_name = "REF")]
    refs: Vec<String>,

    /// Also find and report any tags that point to any identified digest
    #[clap(long)]
    tags: bool,
}

impl CmdInfo {
    pub async fn run(&mut self, verbosity: usize, config: &spfs::Config) -> spfs::Result<i32> {
        let repo = spfs::config::open_repository_from_string(config, self.remote.as_ref()).await?;

        if self.refs.is_empty() {
            self.print_global_info(&repo).await?;
        } else {
            for reference in self.refs.iter() {
                let item = repo.read_ref(reference.as_str()).await?;
                self.pretty_print_ref(item, &repo, verbosity).await?;
            }
        }
        Ok(0)
    }

    async fn format_digest<'repo>(
        &self,
        digest: spfs::encoding::Digest,
        repo: &'repo spfs::storage::RepositoryHandle,
    ) -> spfs::Result<String> {
        if self.tags {
            io::format_digest(digest, DigestFormat::ShortenedWithTags(repo)).await
        } else {
            io::format_digest(digest, DigestFormat::Shortened(repo)).await
        }
    }

    async fn pretty_print_ref(
        &self,
        obj: spfs::graph::Object,
        repo: &spfs::storage::RepositoryHandle,
        verbosity: usize,
    ) -> spfs::Result<()> {
        use spfs::graph::Object;
        match obj {
            Object::Platform(obj) => {
                println!("{}", "platform:".green());
                println!(
                    " {} {}",
                    "refs:".bright_blue(),
                    self.format_digest(obj.digest()?, repo).await?
                );
                println!("{}", "stack:".bright_blue());
                for reference in obj.stack {
                    println!("  - {}", self.format_digest(reference, repo).await?);
                }
            }

            Object::Layer(obj) => {
                println!("{}", "layer:".green());
                println!(
                    " {} {}",
                    "refs:".bright_blue(),
                    self.format_digest(obj.digest()?, repo).await?
                );
                println!(
                    " {} {}",
                    "manifest:".bright_blue(),
                    self.format_digest(obj.manifest, repo).await?
                );
            }

            Object::Manifest(obj) => {
                println!("{}", "manifest:".green());
                let max_entries = match verbosity {
                    0 => 10,
                    1 => 50,
                    _ => usize::MAX,
                };
                let mut count = 0;
                for node in obj.unlock().walk_abs("/spfs") {
                    println!(
                        " {:06o} {} {} {}",
                        node.entry.mode,
                        node.entry.kind,
                        self.format_digest(node.entry.object, repo).await?,
                        node.path,
                    );
                    count += 1;
                    if count >= max_entries {
                        println!("{}", "   ...[truncated] use -vv for more".dimmed());
                        break;
                    }
                }
            }

            Object::Blob(obj) => {
                println!("{}", "blob:".green());
                println!(
                    " {} {}",
                    "digest:".bright_blue(),
                    self.format_digest(obj.payload, repo).await?
                );
                println!(
                    " {} {}",
                    "size:".bright_blue(),
                    spfs::io::format_size(obj.size)
                );
            }
            Object::Tree(_) | Object::Mask => println!("{:?}", obj),
        }
        Ok(())
    }

    /// Display the status of the current runtime.
    async fn print_global_info(&self, repo: &spfs::storage::RepositoryHandle) -> spfs::Result<()> {
        let runtime = spfs::active_runtime().await?;

        println!("{}", "Active Runtime:".green());
        println!(" {}: {}", "id:".bright_blue(), runtime.name());
        println!(
            " {}: {}",
            "editable:".bright_blue(),
            runtime.status.editable
        );
        println!("{}", "stack".bright_blue());
        for digest in runtime.status.stack.iter() {
            println!("  - {}", self.format_digest(*digest, repo).await?);
        }
        println!();

        if !runtime.is_dirty() {
            println!("{}", "No Active Changes".red());
        } else {
            println!("{}", "Run 'spfs diff' for active changes".bright_blue());
        }
        Ok(())
    }
}
